internal/databases/postgres/wal_restore_handler.go

```diff
@@ -0,0 +1,242 @@
+package postgres
+
+import (
+	"errors"
+	"fmt"
+	"github.com/wal-g/wal-g/utility"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/wal-g/wal-g/internal"
+
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+)
+
+type TimelineWithSegmentNo struct {
+	timeline  uint32
+	segmentNo uint64
+}
+
+func NewTimelineWithSegmentNo(tl uint32, seg uint64) *TimelineWithSegmentNo {
+	return &TimelineWithSegmentNo{timeline: tl, segmentNo: seg}
+}
+
+func NewTimelineWithSegmentNoBy(record *TimelineHistoryRecord) *TimelineWithSegmentNo {
+	return NewTimelineWithSegmentNo(record.timeline, getSegmentNoFromLsn(record.lsn))
+}
+
+func getSegmentNoFromLsn(lsn uint64) uint64 {
+	return lsn / WalSegmentSize
```

Comment:
Hm, I guess that this logic may exist somewhere in WAL-G already, are you sure it is not a duplicate?

Reply:
This logic used in WalSegmentNo. Moved this function.